### PR TITLE
feat!: v6.0.0

### LINKED ISSUE #1438: 'Bad state: Cannot add new events after calling close' when using transformEvents
### Description

I get `Bad state: Cannot add new events after calling close (...)` on tests that use `transformEvents`.

### To reproduce

Run the code below. See how the second test fails. Now comment whole `transformEvents` overriden method and run tests again. They pass.

```yaml
dependencies:
  flutter_bloc: ^4.0.1
  rxdart: ^0.23.1
  bloc_test: ^5.1.0
```

```dart
import 'package:bloc_test/bloc_test.dart';
import 'package:flutter_bloc/flutter_bloc.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:rxdart/rxdart.dart';

class TestBloc extends Bloc<bool, bool> {
  @override
  bool get initialState => false;

  @override
  Stream<Transition<bool, bool>> transformEvents(
    Stream<bool> events,
    TransitionFunction<bool, bool> transitionFn,
  ) {
    return events
        .debounceTime(const Duration(milliseconds: 300))
        .switchMap(transitionFn);
  }

  @override
  Stream<bool> mapEventToState(bool event) async* {
    yield event;
  }
}

void main() {
  group('TestBloc', () {
    TestBloc bloc;
    setUp(() => bloc = TestBloc());
    tearDown(() => bloc.close());

    blocTest<TestBloc, bool, bool>(
      'has correct initial state',
      build: () async => bloc,
      skip: 0,
      expect: <bool>[false],
    );

    blocTest<TestBloc, bool, bool>(
      'emits true on event true',
      build: () async => bloc,
      act: (bloc) async => bloc.add(true),
      wait: const Duration(milliseconds: 300),
      expect: <bool>[true],
    );
  });
}

```

### Expected behavior

Both tests pass.

### Logs

```
[✓] Flutter (Channel stable, v1.17.4, on Mac OS X 10.15.5 19F101, locale pl)
    • Flutter version 1.17.4 at /Users/albert/Flutter
    • Framework revision 1ad9baa8b9 (3 weeks ago), 2020-06-17 14:41:16 -0700
    • Engine revision ee76268252
    • Dart version 2.8.4
```

### PR DESCRIPTION
**Pull request by felangel**
## Status

**IN DEVELOPMENT**

**TODOS**
- [x] Merge `package:cubit` into `package:bloc`
  - [x] Update bloc `README` and `example` to include `cubit`
  - [x] Update Cubit Diagram in `README`
- [x] Merge `package:angular_cubit` into `package:angular_bloc`
- [x] Merge `package:cubit_test` into `package:bloc_test`
- [x] Merge `package:flutter_cubit` into `package:flutter_bloc`
  - [x] Update `flutter_bloc` `README` to include `cubit`
  - [x] Update `flutter_bloc` example application to include `cubit`
- [X] Merge `package:hydrated_cubit` into `package:hydrated_bloc`
  - [x] Update `hydrated_bloc` `README` to include `cubit`
  - [x] Update `hydrated_bloc` example application to include `cubit`

## Breaking Changes

YES

## Description

Related to #1427 

### package:bloc

- **BREAKING**: do not emit current state on subscription
- **BREAKING**: `onError` in `BlocObserver` takes a `Cubit` as first parameter
- **BREAKING**: allow blocs and cubits to emit the initial state (@orsenkucher)
- feat: include `cubit` and remove external dependency on [package:cubit](https://pub.dev/packages/cubit)
  - exports class `Cubit`
  - exports class `Change` (`Transition` for `Cubit`)
- feat: `onChange` added to `BlocObserver`
- refactor: apply additional lint rules
- docs: fix inline documentation references

### package:bloc_test

- **BREAKING**: upgrade to `bloc ^6.0.0-dev.1`
- **BREAKING**: `MockBloc` only requires `State` type
- **BREAKING**: `whenListen` only requires `State` type
- **BREAKING**: `blocTest` only requires `State` type
- **BREAKING**: `blocTest` `skip` defaults to `0`
- **BREAKING**: `blocTest` make `build` synchronous
- fix: `blocTest` improve `wait` behavior when debouncing, etc... (closes #1438)
- feat: `blocTest` do not require `async` on `act` and `verify`
- feat: remove external dependency on [package:cubit_test](https://pub.dev/packages/cubit_test)
- feat: `MockBloc` is compatible with `cubit`
- feat: `whenListen` is compatible with `cubit`
- feat: `blocTest` is compatible with `cubit`

### package:flutter_bloc

- **BREAKING**: upgrade to `bloc ^6.0.0-dev.1`
- **BREAKING**: `BlocBuilder` interop with `cubit` (`bloc` parameter renamed to `cubit`)
- **BREAKING**: `BlocListener` interop with `cubit` (`bloc` parameter renamed to `cubit`)
- **BREAKING**: `BlocProvider` interop with `cubit` (`bloc` parameter renamed to `cubit`)
- feat: remove external dependency on [package:flutter_cubit](https://pub.dev/packages/flutter_cubit)
- docs: inline documentation updates
- docs: README updates
- docs: example application updates

### package:hydrated_bloc

- **BREAKING**: upgrade to `bloc ^6.0.0-dev.1`
- fix: json (de)serialization errors ([@orsenkucher](https://github.com/orsenkucher))
  - Hydrated: type `'_InternalLinkedHashMap<dynamic, dynamic>'` is not a subtype of type `'Map<String, dynamic>'` ([#1452](https://github.com/felangel/bloc/issues/1452))
  - Hydrated: HiveError: Cannot write, unknown type: Plan ([#1453](https://github.com/felangel/bloc/issues/1453))
- feat: `StorageNotFound` error thrown if no `Storage` is provided.
- feat: `HydratedCubit` added for `Cubit` interoperability
- feat: `HydratedMixin` added for additional flexibility
- feat: remove external dependency on [package:hydrated_cubit](https://pub.dev/packages/hydrated_cubit)
- docs: inline documentation updates
- docs: README updates
- docs: example application updates

## Type of Change

- [X] ✨ New feature (non-breaking change which adds functionality)
- [x] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [X] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [X] 🧹 Code refactor
- [x] ✅ Build configuration change
- [X] 📝 Documentation
- [x] 🗑️ Chore

### CODE REVIEW

**Comment by orsenkucher** [comment]
Does this mean bloc now will run `distinct` filter only for newly emitted states? (skipping seeded one) @felangel

**Comment by felangel** [comment]
> > **BREAKING:** do not emit current state on subscription
> 
> Does this mean bloc now will run `distinct` filter only for newly emitted states? (skipping seeded one) @felangel

No haha. It means that when you subscribe to the bloc you will not immediately receive the latest state of the bloc. You will only receive states that are emitted after the subscription.

**Comment by orsenkucher** [comment]
hmm, okay, this means if I previously had state: `1`, then subscribed and emitted new `1`, I won't receive it in a BlocListener?
and will never know there was a `1`? @felangel

**Comment by orsenkucher** [comment]
But actually, now | is true for subscription, not only BlocListener

**Comment by felangel** [comment]
> ^
> | But actually, this is now true for subscription, not only BlocListener

yup exactly, it will make the developer experience a lot more consistent and further align with how Dart Streams work.

**Comment by orsenkucher** [comment]
ahah, then I want to point, that from entire set of possible states user can ever see, we eliminate one, causing discontinuity in this "state function" @felangel

**Comment by orsenkucher** [comment]
And from now, fixing it doesn't require breaking change

**Comment by orsenkucher** [comment]
Consider this PR #1433 before merging hydrated cubit to hydrated bloc, to avoid redirecting it one more time 😂

**Comment by orsenkucher** [comment]
LGTM so far!

**Comment by felangel** [comment]
Now that this is getting closer to being ready for a dev release I would love some more 👀  on this.
cc @samandmoore @rodydavis @ResoDev

**Comment by felangel** [comment]
> Approving changes from bed xD
> init -> hydrate is a good one

Thanks so much for your help! Have a good night 💤

**Comment by samandmoore** [comment]
Will take a close look tomorrow! Exciting stuff.

**Comment by irvine5k** [comment]
It sounds very good, go for it :D

**Comment by aytunch** [comment]
Just like @orsenkucher, I am interested in this change too: "BREAKING: do not emit current state on subscription"

Why do we take valuable info away from the hands of the developer. At least leave the default way like you implemented now but give the option to read the last emitted state at the time of subscription if needed by the dev. 

Would this be ok?

**Comment by orsenkucher** [comment]
@aytunch I guess `cubit.state` or `bloc.state` is what you are looking for. It holds current value, so you can access initial state before making subscription

**Comment by orsenkucher** [comment]
BlocBuilder actually uses same trick to draw UI from time of it's creation

**Comment by orestesgaolin** [comment]
Hey, the update/merge looks great 🎉

Looking at the readme I realized that there's no mention of `BlocBuilder`, `BlocListener` and `BlocConsumer`. On the one hand they are described in the docs, but on the other I think it would be good just to mention them and that they can be used both with cubits and blocs.

**Comment by felangel** [comment]
> Hey, the update/merge looks great 🎉
> 
> Looking at the readme I realized that there's no mention of `BlocBuilder`, `BlocListener` and `BlocConsumer`. On the one hand they are described in the docs, but on the other I think it would be good just to mention them and that they can be used both with cubits and blocs.

Thanks for reviewing! 🙏  
Good call, will update 👍

**Comment by samandmoore** [comment]
changes look good to me! will test our a dev release whenever you have time to make one 👍

**Comment by felangel** [comment]
Just published v6.0.0-dev.1 🎉

**Comment by rodydavis** [comment]
Is replay_bloc and replay_cubit missing from the readme? https://pub.dev/packages/replay_bloc

**Comment by felangel** [comment]
> Is replay_bloc and replay_cubit missing from the readme? https://pub.dev/packages/replay_bloc

That package is deprecated and will be part of replay_bloc soon 👍

**Comment by felangel** [comment]
Just published hydrated_bloc v6.0.0-dev.2 which includes:

- fix: json (de)serialization errors ([@orsenkucher](https://github.com/orsenkucher))
  - Hydrated: type `'_InternalLinkedHashMap<dynamic, dynamic>'` is not a subtype of type `'Map<String, dynamic>'` ([#1452](https://github.com/felangel/bloc/issues/1452))
  - Hydrated: HiveError: Cannot write, unknown type: Plan ([#1453](https://github.com/felangel/bloc/issues/1453))

**Comment by felangel** [comment]
Just published `hydrated_bloc v6.0.0-dev.3` which includes:

- fix: complex list (de)serialization errors (thanks [@orsenkucher](https://github.com/orsenkucher))

**Comment by felangel** [comment]
Published bloc_test v7.0.0-dev.2 which includes:

- **BREAKING**: `blocTest` make `build` synchronous
- fix: `blocTest` improve `wait` behavior when debouncing, etc...
- feat: `blocTest` do not require `async` on `act` and `verify`

**Comment by felangel** [comment]
Published hydrated_bloc v6.0.0-dev.4 which includes:

- fix: additional complex list (de)serialization errors ([@orsenkucher](https://github.com/orsenkucher))

File: packages/bloc/lib/src/cubit.dart [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+  /// {@template emit}
+  /// Updates the [state] to the provided [state].
+  /// [emit] does nothing if the [Cubit] has been closed or if the
+  /// [state] being emitted is equal to the current [state].
+  ///
+  /// To allow for the possibility of notifying listeners of the initial state,
+  /// emitting a state which is equal to the initial state is allowed as long
+  /// as it is the first thing emitted by the [Cubit].
+  /// {@endtemplate}
+  @protected
```
Comment (ResoDev) [inline packages/bloc/lib/src/cubit.dart:76]: Should include the `@visibleForTesting` annotation just like the override inside `Bloc`.

Comment (felangel) [reply packages/bloc/lib/src/cubit.dart:76]: No because then you’d have no way to trigger state changes in your cubits 😄

**Comment by felangel** [comment]
Published `hydrated_bloc 6.0.0-dev.5` which includes:
- fix: handle empty case for list traversal

**Comment by orsenkucher** [comment]
> Published `hydrated_bloc 6.0.0-dev.5` which includes:
> 
> * fix: handle empty case for list traversal
@felangel 

Already tested it from git, no errors were detected👌
Will continue to use it further

**Comment by orsenkucher** [comment]
May we also mention in Changelog for `6.0.0` that `Map<String, dynamic>.from()` is unnecessary now and call of `toJson` in optional? @felangel

File: packages/bloc/lib/src/cubit.dart (hunk 1 above)
Comment (felangel) [reply packages/bloc/lib/src/cubit.dart:76]: never mind haha I misunderstood your comment and resolved the issue. Thanks for pointing this out! 💯

**Comment by felangel** [comment]
Published bloc v6.0.0-dev.2 which includes:
- fix: add `@visibleForTesting` to `emit` on class `Cubit`

**Pull request merged**